### PR TITLE
useFocusOutside: rewrite hook to TypeScript, rewrite tests to model RTL and user-event

### LINKED ISSUE #40912: Components: Clicking on iframe in the content close the modal
### Description

I'm trying to load a tinymce editor in a gutenberg modal. I'm using the oficial component: https://github.com/tinymce/tinymce-react#readme

The component load correctly the editor, however clicking or typing something will close the modal.

### Step-by-step reproduction instructions

1. Create a custom component as the one presented here: https://www.tiny.cloud/docs/integrations/react/
2. Create a simple basic modal and put the component there
3. Try to type some texts

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ciampo**
## What?

While trying to fix #40912 I got to rewrite the hook to TypeScript and the tests to modern RTL techniques (including `user-event`)

## Why?

Better tests and TypeScript as a more solid foundation for working on a tricky problem (which I haven't solved, but I thought I'd push these improvements anyway)

## How?

Tests:
- using `user-event` instead of `fireEvent`(or even imperative events called on each element) where possible
- using jest mocking APIs for `document.hasFocus`
- removed useless `TestComponent` wrapper

Hook:
- no runtime changes
- used more specific types for event listeners (e.g. `FocusEventHandler` etc)
- updated the exported type in `useDialog`

## Testing Instructions

Check the `Modal` Storybook example (and modals around the codebase) and make sure that they behave as currently on `trunk`

Code changes can reviewed more easily when approached commit-by-commit

### CODE REVIEW

**Comment by stokesman** [comment]
This looks good. I appreciate the improvements to the unit tests. Though it looks like these tests need more scrutiny because I can try breaking them by removing the hook and 4/5 still pass.
image

 I came across it accidentally while working on a fix for #40912 on another branch. The screenshot is from testing this branch. I'm finding it a bit hard to believe so it'd be great if anyone can try and reproduce it. Not sure if we'd have to address it with this PR because I think it's the same on trunk.

UPDATE:
I reexamined this and I suppose I overreacted. These tests are probably fine 😆. The four that pass without the hook are testing that it doesn't act when it should not. I don't know that there's really any better way of doing it.

File: packages/compose/CHANGELOG.md
Comment (tyxla) [inline packages/compose/CHANGELOG.md:7]: Thanks for adding the missing PR link 👍

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (tyxla) [inline packages/compose/src/hooks/use-focus-outside/test/index.js]: Minor spacing nit:

```suggestion
		// handler is queued
```

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (tyxla) [inline packages/compose/src/hooks/use-focus-outside/index.ts:30]: Could this also use inline docs?

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (tyxla) [inline packages/compose/src/hooks/use-focus-outside/index.ts:41]: Missing the param type?

```suggestion
 * @param  {EventTarget} eventTarget The target from a mouse or touch event.
```

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (tyxla) [inline packages/compose/src/hooks/use-focus-outside/index.ts]: Missing the return type? Also a language nit.

```suggestion
 * @return {boolean} Whether the element is a button.
```

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (tyxla) [inline packages/compose/src/hooks/use-focus-outside/test/index.js]: I know this is pre-existing, but I've found [`unmount`](https://testing-library.com/docs/react-testing-library/api/#unmount) to be a more explicit way to express what we're doing here.

**Review by tyxla — APPROVED** [review]
This looks awesome 💯 

Left a few nits and minor suggestions, but I think it's good to go already 🚀

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/index.ts:41]: In the past, I've usually avoided having types in JSDocs in TypeScript files, as the code itself is already typed. The only exceptions have been for when the JSDoc types where necessary for automatic doc generation.

Adding types here would also require us to add back a lot of JSDoc extra type def (see the code deletions in this file)

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/test/index.js]: Interesting! I'll look into it and report back :)

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (stokesman) [inline packages/compose/src/hooks/use-focus-outside/test/index.js]: I think this could read easier as something like:
```suggestion
	it( "should call handler if focus shifts to element outside component", async () => {
```

Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/test/index.js]: Good point! Done in https://github.com/WordPress/gutenberg/pull/45317/commits/ba97cff2d5d163304ff46d3b1faa32843b106bb2

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/test/index.js]: Done in https://github.com/WordPress/gutenberg/pull/45317/commits/6309e63206c8b1defcc258e13f316a91c0a75a8f

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/index.ts]: Wording improved in https://github.com/WordPress/gutenberg/pull/45317/commits/c9cade699d80454f64fc8ed26237601163df3bf8. For the return type, we can keep https://github.com/WordPress/gutenberg/pull/45317#discussion_r1007787050 open to discuss further in case.

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/index.ts:30]: Done in https://github.com/WordPress/gutenberg/pull/45317/commits/9f5884e22ff85e8c5f99ca06870c962482694339

File: packages/compose/src/hooks/use-focus-outside/test/index.js
Comment (ciampo) [reply packages/compose/src/hooks/use-focus-outside/test/index.js]: Done in https://github.com/WordPress/gutenberg/pull/45317/commits/b4b5737e7f314b030c6ce90f36067cf960710a05

**Comment by ciampo** [comment]
Thank you both @tyxla and @stokesman for the reviews! I've addressed all minor feedback, but I'll wait before merging until we have an agreement on https://github.com/WordPress/gutenberg/pull/45317#discussion_r1007787050.

> I came across it accidentally while working on a fix for https://github.com/WordPress/gutenberg/issues/40912 on another branch.

Interesting, would be curious to see what you're approach is about. I've opened a [draft PR](https://github.com/WordPress/gutenberg/pull/45349) but I'm not sure it's a good approach.

File: packages/compose/src/hooks/use-focus-outside/index.ts
Comment (tyxla) [reply packages/compose/src/hooks/use-focus-outside/index.ts:41]: I generally agree, just wanted to pick your thoughts 👍 Happy to keep it as you suggested.

**Comment by tyxla** [comment]
> I'll wait before merging until we have an agreement on [#45317 (comment)](https://github.com/WordPress/gutenberg/pull/45317#discussion_r1007787050).

I think we agree there already, so you have my green light ✅

**Pull request merged**
